src/main/rx/crsf.c

```diff
@@ -376,7 +376,9 @@ void crsfRxSendTelemetryData(void)
 
 static timeDelta_t crsfFrameDelta(void)
 {
-    return lastRcFrameDelta;
+    const timeDelta_t result = lastRcFrameDelta;
+    lastRcFrameDelta = 0;
+    return result;
```

Comment:
The problem with this is that we will still only be getting deltas between receiving the input data that completed the RC frame and the previous chunk of input data (that might have been part of the same input frame, or a non-RC frame.

I think what we want is deltas between input data chunks that complete RC frames (i.e. if you add them up they will equal the time elapsed).

How about

```
static timeUs_t lastRcFrameUs = 0;

const timeDelta_t result = cmpTimeUs(lastInputDataUs, lastRcFrameUs); // lastInputDataUs set in ISR
lastRcFrameUs = lastInputdataUs;

return result;
```

This can also be implemented centralised in `rx.c`, and only needs `lastInputDataUs` to be `volatile`.

Reply:
Setting aside the EXTI aspect for SPI RX (which is a little different), I don't see this as a problem. The calculation of the `lastFrameDelta` is tied to receiving a valid RX frame (of the proper type) and this also sets the flag that the status function uses to drive subsequent processing. So there isn't any chance of measuring a delta for a partial frame. 

For example with CRSF in the ISR we capture the frame delta and set `crsfFrameDone` once we've reached the end of a packet, the CRC is good, it's for the correct target, and correct packet type.
```
                    case CRSF_FRAMETYPE_RC_CHANNELS_PACKED:
                        if (crsfFrame.frame.deviceAddress == CRSF_ADDRESS_FLIGHT_CONTROLLER) {
                            lastRcFrameDelta = cmpTimeUs(currentTimeUs, lastRcFrameCompleteTimeUs);
                            lastRcFrameCompleteTimeUs = currentTimeUs;
                            crsfFrameDone = true;
                            memcpy(&crsfChannelDataFrame, &crsfFrame, sizeof(crsfFrame));
                        }
                        break;
```
The status function checks `crsfFrameDone` and then processes the channels and returns `RX_FRAME_COMPLETE`.

The returning of `RX_FRAME_COMPLETE` from the status function then triggers the RX task to run and this is where it reads the last captured frame delta and then resets it to 0. The `lastRcFrameCompleteTimeUs` for capturing the next delta is still valid when we received the previous complete frame. So the only way we can call `rxGetFrameDelta()` and read and clear the last frame delta is once we've received a valid RX channels frame.

The other protocols work similarly but are generally simpler because there isn't the need for the multiple frame types, etc. But the same idea in that the ISR detects the end of frame, calculates a delta, then sets a "done" flag. The status function works of the "done" flag to return `RX_FRAME_COMPLETE` and drive the RX task.

Reply:
Yeah, CRSF is a little bit out there as in it does way more than what should be done in the ISR. Most other implementations are better and do the stuff that is not immediately needed to get the data out of the UART buffer in the task check / task function, where it should be done. This means they will not  know if a frame is an RC frame, or if a frame is complete in the ISR. In cases like FPort or Spektrum or FlySky where RC data and telemetry are exchanged over the same UART this will leave us measuring arbitrary timings with not much value - I don't see why we should not do this in a proper way just because CRSF is implemented poorly.

Reply:
I guess from a higher level perspective my point is that we should try to build this in a way that it is as little dependent on the RX protocol implementation as possible. And having the implementation simply return a timestamp of the last time data was received, and then using the 'RC frame received' signalling at the top end of the protocol layer to drive the logic does this.

Reply:
The problem is that what we care about is the delta - not the timestamp of the last valid frame receipt. To calculate a delta upstream we'd then have to cache the last timestamp and do the delta calculation there. So ultimately it's the same just done in a different place.

But there's one additional problem and the key thing that this PR is addressing. We need a way to tell that protocol hasn't received a valid frame. The RX task is going to at a minimum every 30ms even if there are no valid frames from the receiver. So we can't just keep the "last valid frame time" otherwise the frame delta calculation will just increase indefinitely. So we'll need to do the same kind of logic like setting the "lastFrameTime" to 0 and not calculating a delta (returning false) so the upstream logic falls back to the task interval calculation. So in the end it's all the same logically. I guess the only benefit is removing last timestamp and `cmpTimeUs` in each of the ISR routines for a centralized calculation in rx.c. But the core aspect of this PR - the reset of the "last value" after reading - is still needed.

Reply:
@etracer65:

> otherwise the frame delta calculation will just increase indefinitely

I think this is exactly what we should be doing, as it best reflects the reality of what is happening if we want to collect statistics on when RC frames are completed 'on the wire' for cases where the data stream is interrupted. All we need to do is cap the debug value output so it does not overflow.

If we want to / need to collect statistics on task interval based calculation of 'pseudo frames' even for RX protocols that support collection of 'on the wire' statistics then we should do this in a separate debug data stream.

Reply:
No, we don't want to delta to continue to increase. This is not about debugging - that's only a secondary use. The inter-frame delta is used to calculate the RC smoothing filters and used in interpolated feedforward. Having it behave the way your suggesting would break those use cases.  For example, RC smoothing would get into an infinite retraining cycle.

The goal of determining that we are not receiving frames and falling back to the RX task calculation restores functionality that matches how it behaved prior to the addition of low-level calculation in the protocol ISR. It was actually a bug not to consider this case.

Whether we do it as in this PR or switch to a method that captures the last frame time and shifts the calculation of the delta to a higher level as discussed is secondary to needing to differentiate between receiving actual frames and then using the low-level measurement vs. falling back to the high-level RX task calc.

Reply:
@etracer65:

> This is not about debugging

If this is not about debugging then we should not use the `debug` facility for this - users will expect debug values to be actual statistics and not some values that are made up top suit something that isn't debugging.

> The goal of determining that we are not receiving frames and falling back to the RX task calculation restores functionality that matches how it behaved prior to the addition of low-level calculation in the protocol ISR. It was actually a bug not to consider this case.

If it had been obvious that this was intended to be such a hack from the start I would have opposed the addition of it from the start. If we want to actually restore functionality, and not just turn this into a series of made up values that don't help with debugging because the user will never know what they are actually representing then we should restore how this debug value is calculated from task timings, and move reporting of interrupt based timings to a new value.

> Whether we do it as in this PR or switch to a method that captures the last frame time and shifts the calculation of the delta to a higher level as discussed is secondary to needing to differentiate between receiving actual frames and then using the low-level measurement vs. falling back to the high-level RX task calc.

Reply:
The debugging was originally added to support "debugging" the behavior of the sampling and retraining of the RC smoothing cutoffs based on the frame intervals from the receiver. It secondarily has been useful to see previously unknown behaviors from the protocols - particularly recently CRSF. I'm not sure how improving the calculation by moving it into the protocol layer is a "hack". It solves many long term problems caused by having it in the RX task subject to loop time jitter and delays.

The "fallback" logic is there to catch cases where 1. the protocol doesn't support the more accurate measurement; 2. we haven't implemented it yet for a given protocol; 3. we have a state where measurements are possible - like with frame loss. So we want to use the better measurement method but we reserve the ability to fallback to the previous method if needed. So things are better in most cases, but certainly no worse in any case.

Since the debugging that we're looking at is related to RC smoothing, the context of the value is in relationship to that. We have no use for an "interval since the last valid frame". If at some point that becomes important then it could be added without much effort. But `currentRxRefreshRate` isn't it.

Reply:
@etracer65:

> The "fallback" logic is there to catch cases where 1. the protocol doesn't support the more accurate measurement; 2. we haven't implemented it yet for a given protocol; 3. we have a state where measurements are possible - like with frame loss. 

Ok. I fully agree with 1. and 2.. However, with 3, the reality is that **we do not get new RC data while frames are lost**. The fact that the RX stack is re-iterating on processing of data already received to meet timing requirements of some tasks does not change any of this. So I do not see why we should 'fall back' to something that in this case we know is arbitrary task timings not related to RC data changes.

My other point is that with what you are proposing, you are completely neglecting that a lot of the now popular protocols use a single UART for RC data and (bidirectional) telemetry - simply measuring the time difference between an RC frame on the wire and the immediately previous frame (potentially a telemetry frame) will return bogus numbers that are of no use.

Reply:
> However, with 3, the reality is that we do not get new RC data while frames are lost. The fact that the RX stack is re-iterating on processing of data already received to meet timing requirements of some tasks does not change any of this. So I do not see why we should 'fall back' to something that in this case we know is arbitrary task timings not related to RC data changes.

I don't disagree with this. But this is really a bigger discussion about how the fallback logic for the RX task works. The RX task having a maximum interval (30ms) that will cause it to fire in the absence of RX frames driving the on-demand aspect - which if anything is the part that's a bit of a hack. So I'm not saying this can't or shouldn't be improved. But it's outside the scope of this PR.

> My other point is that with what you are proposing, you are completely neglecting that a lot of the now popular protocols use a single UART for RC data and (bidirectional) telemetry - simply measuring the time difference between an RC frame on the wire and the immediately previous frame (potentially a telemetry frame) will return bogus numbers that are of no use.

This is not true. In all the protocols implemented so far the timing is captured contingent on the frame being the RX channels data. For some protocols this is a non-issue as they only have one frame type (SBUS, IBUS, SUMD, etc.). For protocols with multiple frame types the logic only captures the timing for the correct type.

Reply:
> I don't disagree with this. But this is really a bigger discussion about how the fallback logic for the RX task works. The RX task having a maximum interval (30ms) that will cause it to fire in the absence of RX frames driving the on-demand aspect - which if anything is the part that's a bit of a hack. So I'm not saying this can't or shouldn't be improved. But it's outside the scope of this PR.

So then I propose to leave the debug logging as it is until this is cleaned up - adding more hacks and training users that the debug values they get are bogus isn't going to improve the situation.

> This is not true. In all the protocols implemented so far the timing is captured contingent on the frame being the RX channels data. For some protocols this is a non-issue as they only have one frame type (SBUS, IBUS, SUMD, etc.). For protocols with multiple frame types the logic only captures the timing for the correct type.

This is simply wrong. At least for FPort the ISR does not make a distinction between telemetry frames and RC frames when it comes to timing. If all other RC protocols are implemented so poorly that logic to recognise and tread different frame types differently is running at ISR priority then this should be fixed, and not cemented by implementing higher level functionality that requires this to be the case.